### PR TITLE
cli: remove newline insertion after template expansion

### PR DESCRIPTION
**Pull request by higgsd**
In graph mode, a newline is appended whether or not there is content (diffs, file listings, op changes) following the expanded template.  This does not happen in no-graph mode, causing unwanted concatenation, which is particularly problematic for diff headers.

[several edits later]

After several attempts at addressing this, the approved solution is to remove the newline addition.  Built-in templates are unaffected, but users get exactly what they ask for when using a custom template.

# Checklist

If applicable:

- [x] I have updated `CHANGELOG.md`
- [ ] I have updated the documentation (`README.md`, `docs/`, `demos/`)
- [ ] I have updated the config schema (`cli/src/config-schema.json`)
- [x] I have added/updated tests to cover my changes

### CODE REVIEW

File: cli/src/commands/evolog.rs [hunk 1]
```diff
@@ -210,21 +209,29 @@ pub(crate) fn cmd_evolog(
...
-                        formatter,
+                        ui.new_formatter(&mut buffer).as_mut(),
                         &predecessors,
                         &entry.commit,
                         &EverythingMatcher,
                         width,
                     )
                     .block_on()?;
             }
+            write!(raw_output, "{}", &String::from_utf8_lossy(&buffer))?;
```
Comment (yuja) [inline cli/src/commands/evolog.rs]: just `.write_all(&buffer)`

**Comment by higgsd** [comment]
> > perhaps the default templates should be rewritten without newlines, as templates seem more composable without newlines
> 
> Since `--no-graph` output shouldn't insert newlines implicitly, the builtin templates will have to be terminated by newline. 

I see the primary use-case of `--no-graph` as rendering disjoint nodes without a graph, and the ability to produce horizontal output from multiple commits has rather limited utility.  The goal of my PR was just to improve the existing rendering - changing the builtin templates or forcing people to update their custom templates is a tall order.  As such, anything beyond this would be a different FR/PR and I am not sufficiently opinionated to iterate further at this time.

> BTW, someone asked inserting blank lines before/after the `--patch` output. I don't know whether this would be better, but it will mitigate bad formatting of patch contents.

The builtin diff rendering is good at conserving vertical whitespace, and I feel like extra newlines wouldn't be very popular. It would address the problem though.

All just my 2c, though.

**Comment by yuja** [comment]
> I see the primary use-case of `--no-graph` as rendering disjoint nodes without a graph, and the ability to produce horizontal output from multiple commits has rather limited utility.

I think `jj log --no-graph -T` would be often used in scripts.

I'm not strongly against this PR, but doing auto padding of `\n` in more places can provide false notion that templates don't have to include terminal newline. I would rather delete auto padding at all since Sapling's graph renderer handles that.

> > BTW, someone asked inserting blank lines before/after the `--patch` output. I don't know whether this would be better, but it will mitigate bad formatting of patch contents.
> 
> The builtin diff rendering is good at conserving vertical whitespace, and I feel like extra newlines wouldn't be very popular.

`git log -p` inserts blank lines when `--pretty` isn't "oneline", for example.

**Comment by higgsd** [comment]
> > I see the primary use-case of `--no-graph` as rendering disjoint nodes without a graph, and the ability to produce horizontal output from multiple commits has rather limited utility.
> 
> I think `jj log --no-graph -T` would be often used in scripts.

Oops, I've used that extensively but neglected to mention it because the primary output format (ATM) is JSON, which is generally whitespace agnostic.

> I'm not strongly against this PR, but doing auto padding of `\n` in more places can provide false notion that templates don't have to include terminal newline. I would rather delete auto padding at all since Sapling's graph renderer handles that.
> 
> `git log -p` inserts blank lines when `--pretty` isn't "oneline", for example.

I'll experiment with prepending a newline before diff output and see what that looks like.  Should there be any special casing of "empty" diff output?

**Comment by yuja** [comment]
> I'll experiment with prepending a newline before diff output and see what that looks like. Should there be any special casing of "empty" diff output?

I personally don't think we'll need special case for empty output, but if needed, maybe we can use `commit.is_empty()`.

BTW, our `builtin_log_detailed` template includes blank line at the end. We might have to inspect the template output to decide whether blank line should be inserted.

**Comment by higgsd** [comment]
I found a few more cases where templates without newlines were not separated from their "content", and added some test cases for them.

Then I removed a bunch of inconsistent newline usage, and re-added unconditional newlines in as few spots as necessary.  Let me know what you think.

Everything looks decent to me except for `builtin_log_detailed` as you suspected; the default `jj show` output ends up with one too many newlines.  Should I trim it?

**Comment by higgsd** [comment]
I don't see an easy solution here.  Without modifying `builtin_log_detailed` ([224f101](https://github.com/jj-vcs/jj/pull/8327/commits/224f101a4080fdab8985e361e6875b3d0703f72f)), certain `jj show` outputs produce doubled newlines.

Removing the trailing newline from `builtin_log_detailed` ([eff91f7](https://github.com/jj-vcs/jj/pull/8327/commits/eff91f7829f359eb13a79575dc88e3e446b00828)) will restore `jj show` behavior but then we lose a line from each `jj log -T builtin_log_detailed` item.  I'm just rearranging deck chairs here and it's becoming tedious.

**Comment by yuja** [comment]
One weird hack we could use is, if template output ends with blank line, insert blank line after the diff output. IOW, copy the template style to diff output. This wouldn't solve the original problem, but maybe that's okay since template should be terminated with newline.

BTW, it's more important to insert blank line _after_ the diff output, not before. Otherwise, the diff output would look more related to the next commit entry.

**Comment by higgsd** [comment]
I am back to my original approach, which was to adjust the existing "newline fixup" behavior.

That is, only insert the newline when (a) the template expansion does not end with a newline and (b) the output will show additional content (even when empty).  The end result keeps 99% of existing newline behavior in tests and day-to-day usage, while improving the corner case(s) that I identified.

I'm not super pleased with all the raw_output/buffer behavior, but don't have the Rust chops (yet?) to make it look nicer.

P.S. Any concern I had about sapling graph node rendering was either a red herring or irrelevant to these changes, so far as I can tell, sorry.

File: cli/src/commands/log.rs [hunk 2]
```diff
@@ -273,15 +272,15 @@ pub(crate) fn cmd_log(
...
                 })?;
-                if !buffer.ends_with(b"\n") {
-                    buffer.push(b'\n');
-                }
                 if let Some(renderer) = &diff_renderer {
-                    let mut formatter = ui.new_formatter(&mut buffer);
+                    // ensure template expansion is separated from any content
+                    if !buffer.ends_with(b"\n") {
+                        buffer.push(b'\n');
+                    }
```
Comment (yuja) [inline cli/src/commands/log.rs]: Can't we simply delete special case for templates that lack newline terminator?

I kind of agree that it's convenient if newline in templates can be omitted so we can write `-Tjson(self)` for example, but the rule should apply globally to non-graph/non-patch outputs. It might be some config flag, but I don't have a good idea.

BTW, the current implementation is a bit buggy. If `buffer` is empty, no newline should be added.

File: cli/src/commands/log.rs [hunk 3]
```diff
@@ -333,14 +332,27 @@ pub(crate) fn cmd_log(
...
                     let width = ui.term_width();
                     renderer
-                        .show_patch(ui, formatter, &commit, matcher.as_ref(), width)
+                        .show_patch(
+                            ui,
+                            ui.new_formatter(&mut buffer).as_mut(),
+                            &commit,
+                            matcher.as_ref(),
+                            width,
+                        )
```
Comment (yuja) [inline cli/src/commands/log.rs]: nit: we can still write the patch output directly to stdout.

Comment (higgsd) [reply cli/src/commands/log.rs]: No longer relevant to latest push

File: cli/src/commands/log.rs (hunk 2 above)
Comment (higgsd) [reply cli/src/commands/log.rs]: Obsolete w/ latest push.  Filing empty buffer comment away in case we circle back (yet) again.

**Comment by higgsd** [comment]
In the latest push, I removed _all_ newline compensation.  If the user didn't use/build the right template, they get exactly what they asked for, mangled content or not.  It doesn't seem to cause problems with built-in templates, but custom templates might need adjustment, so I put it in the breaking changes section of CHANGELOG.md.

**Comment by glehmann** [comment]
The behavior has changed with `--summary`. I think this is linked to this PR: unlike before, the summary begins just after template content:

image

Is it the new expected behavior?

**Comment by higgsd** [comment]
Yes, this is expected as a result of this PR: diffs, --stat, --name-only, etc are all in the same behavioral boat.
